### PR TITLE
[Bots] Cleanup Bot Spell Functions, reduce reliance on NPC Functions/Attributes

### PR DESCRIPTION
**Pull request by Aeadoin**
In Preparation of adding new Spell fields to Bot_Spells_Entries, I figured it would be easiest to separate the Spell functions Bots use.
This will reduce the chance of any issues popping up with NPC casting by splitting these off now.

Eventually I'd like to look into fully splitting off Bots.

### CODE REVIEW

File: zone/bot.cpp [hunk 1]
```diff
@@ -2173,6 +2173,46 @@ bool Bot::Process()
 	return true;
 }
 
+void Bot::AI_Bot_Start(uint32 iMoveDelay) {
+	Mob::AI_Start(iMoveDelay);
+	if (!pAIControlled)
```
Comment (Kinglykrab) [inline zone/bot.cpp:2178]: Brackets here?

File: zone/botspellsai.cpp [hunk 2]
```diff
@@ -2674,4 +2674,337 @@ uint8 Bot::GetChanceToCastBySpellType(uint32 spellType)
...
+
+		if(RuleB(Spells, NPCInnateProcOverride))
+			innate_proc_spell_id = attack_proc_spell;
+	}
+
+	if (IsValidSpell(range_proc_spell))
+		AddRangedProc(range_proc_spell, (rproc_chance + 100));
+
+	if (IsValidSpell(defensive_proc_spell))
+		AddDefensiveProc(defensive_proc_spell, (dproc_chance + 100));
```
Comment (Kinglykrab) [inline zone/botspellsai.cpp:2827]: Brackets here and above.

File: zone/botspellsai.cpp [hunk 3]
```diff
@@ -2674,4 +2674,337 @@ uint8 Bot::GetChanceToCastBySpellType(uint32 spellType)
...
+	AISpellVar.pursue_no_sp_recast_min = (_pursue_no_sp_recast_min) ? _pursue_no_sp_recast_min : RuleI(Spells, AI_PursueNoSpellMinRecast);
+	AISpellVar.pursue_no_sp_recast_max = (_pursue_no_sp_recast_max) ? _pursue_no_sp_recast_max : RuleI(Spells, AI_PursueNoSpellMaxRecast);
+	AISpellVar.pursue_detrimental_chance = (_pursue_detrimental_chance) ? _pursue_detrimental_chance : RuleI(Spells, AI_PursueDetrimentalChance);
+	AISpellVar.idle_no_sp_recast_min = (_idle_no_sp_recast_min) ? _idle_no_sp_recast_min : RuleI(Spells, AI_IdleNoSpellMinRecast);
+	AISpellVar.idle_no_sp_recast_max = (_idle_no_sp_recast_max) ? _idle_no_sp_recast_max : RuleI(Spells, AI_IdleNoSpellMaxRecast);
+	AISpellVar.idle_beneficial_chance = (_idle_beneficial_chance) ? _idle_beneficial_chance : RuleI(Spells, AI_IdleBeneficialChance);
+
+	if (AIBot_spells.empty())
+		AIautocastspell_timer->Disable();
+	else
```
Comment (Kinglykrab) [inline zone/botspellsai.cpp:2846]: Brackets.

File: zone/botspellsai.cpp [hunk 4]
```diff
@@ -2674,4 +2674,337 @@ uint8 Bot::GetChanceToCastBySpellType(uint32 spellType)
...
+
+bool IsSpellInBotList(DBnpcspells_Struct* spell_list, uint16 iSpellID) {
+	auto it = std::find_if(spell_list->entries.begin(), spell_list->entries.end(),
+			       [iSpellID](const DBnpcspells_entries_Struct &a) { return a.spellid == iSpellID; });
+	return it != spell_list->entries.end();
+}
+
+DBnpcspells_Struct *ZoneDatabase::GetBotSpells(uint32 iDBSpellsID)
+{
+	if (iDBSpellsID == 0)
```
Comment (Kinglykrab) [inline zone/botspellsai.cpp:2859]: Brackets.

File: zone/botspellsai.cpp [hunk 5]
```diff
@@ -2674,4 +2674,337 @@ uint8 Bot::GetChanceToCastBySpellType(uint32 spellType)
...
+						 "pursue_no_sp_recast_min, pursue_no_sp_recast_max, "
+						 "pursue_d_chance, idle_no_sp_recast_min, idle_no_sp_recast_max, "
+						 "idle_b_chance FROM npc_spells WHERE id=%d",
+						 iDBSpellsID);
+		auto results = QueryDatabase(query);
+		if (!results.Success()) {
+			return nullptr;
+		}
+
+		if (results.RowCount() != 1)
```
Comment (Kinglykrab) [inline zone/botspellsai.cpp:2884]: Brackets.

File: zone/botspellsai.cpp [hunk 6]
```diff
@@ -2674,4 +2674,337 @@ uint8 Bot::GetChanceToCastBySpellType(uint32 spellType)
...
+			entry.min_hp = atoi(row[7]);
+			entry.max_hp = atoi(row[8]);
+
+			// some spell types don't make much since to be priority 0, so fix that
+			if (!(entry.type & SPELL_TYPES_INNATE) && entry.priority == 0)
+				entry.priority = 1;
+
+			if (row[9])
+				entry.resist_adjust = atoi(row[9]);
+			else if (IsValidSpell(spell_id))
```
Comment (Kinglykrab) [inline zone/botspellsai.cpp:2943]: Brackets.

File: zone/botspellsai.cpp [hunk 7]
```diff
@@ -2674,4 +2674,337 @@ uint8 Bot::GetChanceToCastBySpellType(uint32 spellType)
...
+	return nullptr;
+}
+
+// adds a spell to the list, taking into account priority and resorting list as needed.
+void Bot::AddSpellToBotList(int16 iPriority, uint16 iSpellID, uint32 iType,
+							int16 iManaCost, int32 iRecastDelay, int16 iResistAdjust, int8 min_hp, int8 max_hp)
+{
+
+	if(!IsValidSpell(iSpellID))
+		return;
```
Comment (Kinglykrab) [inline zone/botspellsai.cpp:2963]: Brackets.

File: zone/botspellsai.cpp [hunk 8]
```diff
@@ -2674,4 +2674,337 @@ uint8 Bot::GetChanceToCastBySpellType(uint32 spellType)
...
+	t.recast_delay = iRecastDelay;
+	t.time_cancast = 0;
+	t.resist_adjust = iResistAdjust;
+	t.min_hp = min_hp;
+	t.max_hp = max_hp;
+
+	AIBot_spells.push_back(t);
+
+	// If we're going from an empty list, we need to start the timer
+	if (AIBot_spells.empty())
```
Comment (Kinglykrab) [inline zone/botspellsai.cpp:2981]: Brackets.

File: zone/botspellsai.cpp [hunk 9]
```diff
@@ -2674,4 +2674,337 @@ uint8 Bot::GetChanceToCastBySpellType(uint32 spellType)
...
+							AIBot_spells[casting_spell_AIindex].time_cancast = Timer::GetCurrentTime() + (AIBot_spells[casting_spell_AIindex].recast_delay*1000);
+					}
+					else
+						AIBot_spells[casting_spell_AIindex].time_cancast = Timer::GetCurrentTime() + spells[AIBot_spells[casting_spell_AIindex].spellid].recast_time;
+			}
+			if (recovery_time < AIautocastspell_timer->GetSetAtTrigger())
+				recovery_time = AIautocastspell_timer->GetSetAtTrigger();
+			AIautocastspell_timer->Start(recovery_time, false);
+		}
+		else
```
Comment (Kinglykrab) [inline zone/botspellsai.cpp:3004]: Brackets.

File: zone/zonedb.h [hunk 10]
```diff
@@ -491,6 +491,9 @@ class ZoneDatabase : public SharedDatabase {
 	void ClearNPCSpells() { npc_spells_cache.clear(); npc_spells_loadtried.clear(); }
 	const NPCType* LoadNPCTypesData(uint32 id, bool bulk_load = false);
 
+	/*Bots	*/
+	DBnpcspells_Struct*				GetBotSpells(uint32 iDBSpellsID);
```
Comment (Kinglykrab) [inline zone/zonedb.h:495]: Formatting.

File: zone/npc.cpp [hunk 11]
```diff
@@ -309,8 +314,16 @@ NPC::NPC(const NPCType *npc_type_data, Spawn2 *in_respawn, const glm::vec4 &posi
 	AISpellVar.idle_no_sp_recast_max           = static_cast<uint32>(RuleI(Spells, AI_IdleNoSpellMaxRecast));
 	AISpellVar.idle_beneficial_chance          = static_cast<uint8> (RuleI(Spells, AI_IdleBeneficialChance));
 
-	AI_Init();
-	AI_Start();
+	// It's possible for IsBot() to not be set yet during Bot loading, so have to use an alternative to catch Bots
+	if (!EQ::ValueWithin(npc_type_data->npc_spells_id, 3001, 3016)) {
```
Comment (Kinglykrab) [inline zone/npc.cpp:318]: Constants? BotSpellIDs::Warrior for example.

File: zone/botspellsai.cpp [hunk 12]
```diff
@@ -2674,4 +2674,337 @@ uint8 Bot::GetChanceToCastBySpellType(uint32 spellType)
...
+		}
+	}
+	LogAI("fin (parent list)");
+	if (spell_list) {
+		for (const auto &iter : spell_list->entries) {
+			LogAI("([{}]) [{}]", iter.spellid, spells[iter.spellid].name);
+		}
+	}
+	LogAI("fin (spell list)");
+#endif
```
Comment (Kinglykrab) [inline zone/botspellsai.cpp:2722]: Not sure if we need this debug stuff.

Comment (Aeadoin) [reply zone/botspellsai.cpp:2722]: Was a direct copy from `bool NPC::AI_AddNPCSpells(uint32 iDBSpellsID)`
I can remove it from Bots if you don't think it has any value.

Comment (Kinglykrab) [reply zone/botspellsai.cpp:2722]: With our new logging stuff you can turn logs off/on, so can just have whatever logs we want by default and if they’re disabled they just don’t show up.

File: zone/botspellsai.cpp [hunk 13]
```diff
@@ -2674,4 +2674,337 @@ uint8 Bot::GetChanceToCastBySpellType(uint32 spellType)
 	return database.botdb.GetSpellCastingChance(spell_type_index, class_index, stance_index, type_index);
 }
 
+bool IsSpellInBotList(DBnpcspells_Struct* spell_list, uint16 iSpellID);
```
[the remark this answers is not in the document]
Comment (Aeadoin) [reply zone/botspellsai.cpp:2677]: Moved declaration to bot.h

File: zone/npc.cpp (hunk 11 above)
Comment (Kinglykrab) [reply zone/npc.cpp:318]: Or BotSpellLists::Warrior. Anything is better than magic numbers.

Comment (Aeadoin) [reply zone/npc.cpp]: Added constants.

**Comment by Aeadoin** [comment]
Found a couple instances where bots could cause a zone crash, and resolved those. tested for an extended length of time with different bot types.

File: zone/bot.cpp [hunk 14]
```diff
@@ -2259,7 +2300,7 @@ void Bot::BotRangedAttack(Mob* other) {
 	if (spellbonuses.NegateIfCombat)
 		BuffFadeByEffect(SE_NegateIfCombat);
 
-	if(hidden || improved_hidden){
+	if(hidden || improved_hidden) {
```
Comment (Kinglykrab) [inline zone/bot.cpp:2303]: Formatting.

File: zone/bot.cpp [hunk 15]
```diff
@@ -6430,9 +6471,9 @@ void Bot::DoClassAttacks(Mob *target, bool IsRiposte) {
 		return;
 	}
 
-	if(ka_time){
+	if(ka_time) {
 
-		switch(GetClass()){
+		switch(GetClass()) {
```
Comment (Kinglykrab) [inline zone/bot.cpp:6476]: Formatting.

File: zone/bot.cpp [hunk 16]
```diff
@@ -6430,9 +6471,9 @@ void Bot::DoClassAttacks(Mob *target, bool IsRiposte) {
 		return;
 	}
 
-	if(ka_time){
+	if(ka_time) {
```
Comment (Kinglykrab) [inline zone/bot.cpp:6474]: Formatting.

File: zone/bot.cpp [hunk 17]
```diff
@@ -6532,7 +6573,7 @@ void Bot::DoClassAttacks(Mob *target, bool IsRiposte) {
 	bool did_attack = false;
 	switch(GetClass()) {
 		case WARRIOR:
-			if(level >= RuleI(Combat, NPCBashKickLevel)){
+			if(level >= RuleI(Combat, NPCBashKickLevel)) {
```
Comment (Kinglykrab) [inline zone/bot.cpp:6576]: Formatting.

File: zone/bot.cpp [hunk 18]
```diff
@@ -6552,7 +6593,7 @@ void Bot::DoClassAttacks(Mob *target, bool IsRiposte) {
 		case CLERIC:
 		case SHADOWKNIGHT:
 		case PALADIN:
-			if(level >= RuleI(Combat, NPCBashKickLevel)){
+			if(level >= RuleI(Combat, NPCBashKickLevel)) {
```
Comment (Kinglykrab) [inline zone/bot.cpp:6596]: Formatting.

File: zone/bot.cpp [hunk 19]
```diff
@@ -7436,16 +7477,16 @@ bool Bot::CastSpell(uint16 spell_id, uint16 target_id, EQ::spells::CastingSlot s
...
 				if(casting_spell_id)
-					AI_Event_SpellCastFinished(false, static_cast<uint16>(casting_spell_slot));
+					AI_Bot_Event_SpellCastFinished(false, static_cast<uint16>(casting_spell_slot));
 
 				return false;
 			}
 		}
 
-		if(IsDetrimentalSpell(spell_id) && !zone->CanDoCombat()){
+		if(IsDetrimentalSpell(spell_id) && !zone->CanDoCombat()) {
```
Comment (Kinglykrab) [inline zone/bot.cpp:7486]: Formatting.

File: zone/bot.cpp [hunk 20]
```diff
@@ -7733,7 +7774,7 @@ bool Bot::DoFinishedSpellSingleTarget(uint16 spell_id, Mob* spellTarget, EQ::spe
 		if(IsGrouped() && (spellTarget->IsBot() || spellTarget->IsClient()) && RuleB(Bots, GroupBuffing)) {
 			bool noGroupSpell = false;
 			uint16 thespell = spell_id;
-			for(int i = 0; i < AIspells.size(); i++) {
+			for(int i = 0; i < AIBot_spells.size(); i++) {
```
Comment (Kinglykrab) [inline zone/bot.cpp:7777]: Formatting.

File: zone/botspellsai.cpp [hunk 21]
```diff
@@ -87,13 +87,14 @@ bool Bot::AICastSpell(Mob* tar, uint8 iChance, uint32 iSpellTypes) {
...
-				if(!addMob){
+				if(!addMob) {
 					//Say("!addMob.");
-					break;}
-
-				if(!(!addMob->IsImmuneToSpell(botSpell.SpellId, this) && addMob->CanBuffStack(botSpell.SpellId, botLevel, true) >= 0))
 					break;
+				}
 
+				if(!(!addMob->IsImmuneToSpell(botSpell.SpellId, this) && addMob->CanBuffStack(botSpell.SpellId, botLevel, true) >= 0)) {
```
Comment (Kinglykrab) [inline zone/botspellsai.cpp:95]: Formatting.

File: zone/botspellsai.cpp [hunk 22]
```diff
@@ -87,13 +87,14 @@ bool Bot::AICastSpell(Mob* tar, uint8 iChance, uint32 iSpellTypes) {
 
 				Mob* addMob = GetFirstIncomingMobToMez(this, botSpell);
 
-				if(!addMob){
+				if(!addMob) {
```
Comment (Kinglykrab) [inline zone/botspellsai.cpp:90]: Formatting.

File: zone/botspellsai.cpp [hunk 23]
```diff
@@ -135,27 +136,30 @@ bool Bot::AICastSpell(Mob* tar, uint8 iChance, uint32 iSpellTypes) {
 						if(hpr < 35) {
 							botSpell = GetBestBotSpellForFastHeal(this);
 						}
-						else if(hpr >= 35 && hpr < 70){
-							if(GetNumberNeedingHealedInGroup(60, false) >= 3)
+						else if(hpr >= 35 && hpr < 70) {
+							if(GetNumberNeedingHealedInGroup(60, false) >= 3) {
```
Comment (Kinglykrab) [inline zone/botspellsai.cpp:140]: Formatting.

File: zone/botspellsai.cpp [hunk 24]
```diff
@@ -135,27 +136,30 @@ bool Bot::AICastSpell(Mob* tar, uint8 iChance, uint32 iSpellTypes) {
...
 								botSpell = GetBestBotSpellForGroupHeal(this);
-
-							if(botSpell.SpellId == 0)
+							}
+							if(botSpell.SpellId == 0) {
```
Comment (Kinglykrab) [inline zone/botspellsai.cpp:143]: Formatting.

File: zone/botspellsai.cpp [hunk 25]
```diff
@@ -135,27 +136,30 @@ bool Bot::AICastSpell(Mob* tar, uint8 iChance, uint32 iSpellTypes) {
...
 								botSpell = GetBestBotSpellForPercentageHeal(this);
+							}
 						}
-						else if(hpr >= 70 && hpr < 95){
-							if(GetNumberNeedingHealedInGroup(80, false) >= 3)
+						else if(hpr >= 70 && hpr < 95) {
```
Comment (Kinglykrab) [inline zone/botspellsai.cpp:147]: Formatting.

File: zone/botspellsai.cpp [hunk 26]
```diff
@@ -135,27 +136,30 @@ bool Bot::AICastSpell(Mob* tar, uint8 iChance, uint32 iSpellTypes) {
...
+							if(GetNumberNeedingHealedInGroup(80, false) >= 3) {
 								botSpell = GetBestBotSpellForGroupHealOverTime(this);
-
-							if(hasAggro)
+							}
+							if(hasAggro) {
```
Comment (Kinglykrab) [inline zone/botspellsai.cpp:151]: Formatting.

File: zone/botspellsai.cpp [hunk 27]
```diff
@@ -135,27 +136,30 @@ bool Bot::AICastSpell(Mob* tar, uint8 iChance, uint32 iSpellTypes) {
...
 								botSpell = GetBestBotSpellForPercentageHeal(this);
+							}
 						}
 						else {
-							if(!tar->FindType(SE_HealOverTime))
+							if(!tar->FindType(SE_HealOverTime)) {
```
Comment (Kinglykrab) [inline zone/botspellsai.cpp:156]: Formatting.

File: zone/botspellsai.cpp [hunk 28]
```diff
@@ -135,27 +136,30 @@ bool Bot::AICastSpell(Mob* tar, uint8 iChance, uint32 iSpellTypes) {
...
 								botSpell = GetBestBotSpellForHealOverTime(this);
+							}
 						}
 					}
 					else if ((botClass == CLERIC) || (botClass == DRUID) || (botClass == SHAMAN) || (botClass == PALADIN)) {
-						if(GetNumberNeedingHealedInGroup(40, true) >= 2){
+						if(GetNumberNeedingHealedInGroup(40, true) >= 2) {
```
Comment (Kinglykrab) [inline zone/botspellsai.cpp:162]: Formatting.

File: zone/botspellsai.cpp [hunk 29]
```diff
@@ -223,7 +227,7 @@ bool Bot::AICastSpell(Mob* tar, uint8 iChance, uint32 iSpellTypes) {
 					if(botSpell.SpellId == 0)
 						botSpell = GetFirstBotSpellForSingleTargetHeal(this);
 
-					if(botSpell.SpellId == 0 && botClass == BARD){
+					if(botSpell.SpellId == 0 && botClass == BARD) {
```
Comment (Kinglykrab) [inline zone/botspellsai.cpp:230]: Formatting.

File: zone/botspellsai.cpp [hunk 30]
```diff
@@ -257,7 +261,7 @@ bool Bot::AICastSpell(Mob* tar, uint8 iChance, uint32 iSpellTypes) {
 							}
 						}*/
 						if(botClass != BARD) {
-							if(IsGroupSpell(botSpell.SpellId)){
+							if(IsGroupSpell(botSpell.SpellId)) {
```
Comment (Kinglykrab) [inline zone/botspellsai.cpp:264]: Formatting.

File: zone/botspellsai.cpp [hunk 31]
```diff
@@ -347,7 +351,7 @@ bool Bot::AICastSpell(Mob* tar, uint8 iChance, uint32 iSpellTypes) {
 					// Put the zone levitate and movement check here since bots are able to bypass the client casting check
 					if((IsEffectInSpell(selectedBotSpell.SpellId, SE_Levitate) && !zone->CanLevitate())
 						|| (IsEffectInSpell(selectedBotSpell.SpellId, SE_MovementSpeed) && !zone->CanCastOutdoor())) {
-							if(botClass != BARD || !IsSpellUsableThisZoneType(selectedBotSpell.SpellId, zone->GetZoneType())){
+							if(botClass != BARD || !IsSpellUsableThisZoneType(selectedBotSpell.SpellId, zone->GetZoneType())) {
```
Comment (Kinglykrab) [inline zone/botspellsai.cpp:354]: Formatting.

File: zone/botspellsai.cpp [hunk 32]
```diff
@@ -926,7 +930,7 @@ bool Bot::AICastSpell(Mob* tar, uint8 iChance, uint32 iSpellTypes) {
 
 				if(castedSpell) {
 					if(botClass != BARD) {
-						if(IsGroupSpell(botSpell.SpellId)){
+						if(IsGroupSpell(botSpell.SpellId)) {
```
Comment (Kinglykrab) [inline zone/botspellsai.cpp:933]: Formatting.

File: zone/botspellsai.cpp [hunk 33]
```diff
@@ -1557,7 +1563,7 @@ bool Bot::AIHealRotation(Mob* tar, bool useFastHeals) {
 		if(botSpell.SpellId == 0)
 			botSpell = GetFirstBotSpellForSingleTargetHeal(this);
 
-		if(botSpell.SpellId == 0){
+		if(botSpell.SpellId == 0) {
```
Comment (Kinglykrab) [inline zone/botspellsai.cpp:1566]: Formatting.

File: zone/botspellsai.cpp [hunk 34]
```diff
@@ -2674,4 +2680,344 @@ uint8 Bot::GetChanceToCastBySpellType(uint32 spellType)
...
+	}
+
+	std::sort(AIBot_spells.begin(), AIBot_spells.end(), [](const BotSpells_Struct& a, const BotSpells_Struct& b) {
+		return a.priority > b.priority;
+	});
+
+	if (IsValidSpell(attack_proc_spell)) {
+		AddProcToWeapon(attack_proc_spell, true, proc_chance);
+
+		if(RuleB(Spells, NPCInnateProcOverride)) {
```
Comment (Kinglykrab) [inline zone/botspellsai.cpp:2820]: Formatting.

**Review by Kinglykrab — APPROVED** [review]
Tested code, looks good to me. Just formatting stuff if we care.

**Pull request merged**